src/kinetics/RateFactory.cpp

```
@@ -0,0 +1,102 @@
+// at https://cantera.org/license.txt for license and copyright information.
+
+#include "cantera/kinetics/RateFactory.h"
+#include "cantera/kinetics/MultiRate.h"
+#include "cantera/thermo/ThermoPhase.h"
+#include "cantera/kinetics/Kinetics.h"
+#include "cantera/base/AnyMap.h"
+
+namespace Cantera
+{
+
+RateFactory* RateFactory::s_factory = 0;
+std::mutex RateFactory::rate_mutex;
+
+RateFactory::RateFactory()
+{
+    // ArrheniusRate evaluator
+    reg("ArrheniusRate", [](const AnyMap& node, const Units& rate_units) {
+        return new ArrheniusRate(node, rate_units);
+    });
+    addAlias("ArrheniusRate", "");
+    addAlias("ArrheniusRate", "elementary");
+    addAlias("ArrheniusRate", "arrhenius");
+    addAlias("ArrheniusRate", "three-body");
+    addAlias("ArrheniusRate", "threebody");
+    addAlias("ArrheniusRate", "three_body");
+
+    // PlogRate evaluator
+    reg("PlogRate", [](const AnyMap& node, const Units& rate_units) {
+        return new PlogRate(node, rate_units);
+    });
+    addAlias("PlogRate", "pressure-dependent-Arrhenius");
+    addAlias("PlogRate", "plog");
+    addAlias("PlogRate", "pressure-dependent-Arrhenius");
+
+    // ChebyshevRate evaluator
+    reg("ChebyshevRate", [](const AnyMap& node, const Units& rate_units) {
+        return new ChebyshevRate3(node, rate_units);
+    });
+    addAlias("ChebyshevRate", "Chebyshev");
+    addAlias("ChebyshevRate", "chebyshev");
+
+    // CustomFunc1Rate evaluator
+    reg("custom-function", [](const AnyMap& node, const Units& rate_units) {
+        return new CustomFunc1Rate(node, rate_units);
+    });
+    addAlias("custom-function", "custom-rate-function");
+}
+
+shared_ptr<ReactionRateBase> newRate(const std::string& type)
+{
+    if (RateFactory::factory()->exists(type)) {
+        return shared_ptr<ReactionRateBase> (
+            RateFactory::factory()->create(type, AnyMap(), Units(0.0)));
+    }
+    return shared_ptr<ReactionRateBase> ();
```

This behavior of silently returning a null object when the object type isn't found is a bit surprising, and different from how our other factories work. What's the motivation?

---

Excellent point. It turns out to not be necessary (at least at this point).